go/ai/generate.go

```diff
@@ -86,8 +86,170 @@ func LookupModel(provider, name string) *Model {
+func WithSystemPrompt(prompt string) GenerateParamsBuilder {
+	return func(req *GenerateParams) error {
+		req.SystemPrompt = NewSystemTextMessage(prompt)
+		return nil
+	}
+}
+
+// WithMessages adds provided messages to GenerateRequest.
+func WithMessages(messages ...*Message) GenerateParamsBuilder {
+	return func(req *GenerateParams) error {
+		req.Request.Messages = append(req.Request.Messages, messages...)
+		return nil
+	}
+}
+
+// WithHistory adds provided history messages to the begining of GenerateRequest.Messages.
+func WithHistory(history ...*Message) GenerateParamsBuilder {
+	return func(req *GenerateParams) error {
+		req.History = history
+		return nil
+	}
+}
+
+// WithConfig adds provided config to GenerateRequest.
+func WithConfig(config any) GenerateParamsBuilder {
+	return func(req *GenerateParams) error {
+		req.Request.Config = config
+		return nil
+	}
+}
+
+// WithCandidates adds provided candidate count to GenerateRequest.
+func WithCandidates(c int) GenerateParamsBuilder {
+	return func(req *GenerateParams) error {
+		req.Request.Candidates = c
+		return nil
+	}
+}
+
+// WithContext adds provided context to GenerateRequest.
+func WithContext(c ...any) GenerateParamsBuilder {
+	return func(req *GenerateParams) error {
+		req.Request.Context = c
+		return nil
+	}
+}
+
+// WithTools adds provided tools to GenerateRequest.
+func WithTools(tools ...Tool) GenerateParamsBuilder {
+	return func(req *GenerateParams) error {
+		var toolDefs []*ToolDefinition
+		for _, t := range tools {
+			toolDefs = append(toolDefs, t.Definition())
+		}
+		req.Request.Tools = toolDefs
+		return nil
+	}
+}
+
+// WithOutputSchema adds provided output schema to GenerateRequest.
+func WithOutputSchema(schema any) GenerateParamsBuilder {
+	return func(req *GenerateParams) error {
+		if req.Request.Output != nil && req.Request.Output.Schema != nil {
+			return errors.New("cannot set Request.Output.Schema (WithOutputSchema) more than once")
+		}
+		if req.Request.Output == nil {
+			req.Request.Output = &GenerateRequestOutput{}
+			req.Request.Output.Format = OutputFormatJSON
+		}
+		req.Request.Output.Schema = base.SchemaAsMap(base.InferJSONSchemaNonReferencing(schema))
+		return nil
+	}
+}
+
+// WithOutputFormat adds provided output format to GenerateRequest.
+func WithOutputFormat(format OutputFormat) GenerateParamsBuilder {
+	return func(req *GenerateParams) error {
+		if req.Request.Output == nil {
+			req.Request.Output = &GenerateRequestOutput{}
+		}
+		req.Request.Output.Format = format
+		return nil
+	}
+}
+
+// WithStreaming adds a streaming callback to the generate request.
+func WithStreaming(cb ModelStreamingCallback) GenerateParamsBuilder {
+	return func(req *GenerateParams) error {
+		req.Stream = cb
+		return nil
+	}
+}
+
+// Generate run generate request for this model. Returns GenerateResponse struct.
+func (m *Model) Generate(ctx context.Context, withs ...GenerateParamsBuilder) (*GenerateResponse, error) {
```

Comment:
typically `opts ...GenerateOption`

Reply:
But I am confused: I thought on the discord we were going to make this more low-level, taking a GenerateRequest, and have functions like `ai.GenerateXXX`.

Reply:
I like `GenerateOption`! Renamed.
re: `ai.GenerateXXX` that will be in the next PR. Want to keep these changes manageable/reviewable. In retrospect I should've started with that, but was super excited about this first.